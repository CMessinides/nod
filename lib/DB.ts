import { Pool } from "pg";
import { connectionString } from "../config/db.config";
import { dev } from "../config/server.config";

const pool = new Pool({ connectionString, ssl: !dev });

export default {
	query: pool.query
};
