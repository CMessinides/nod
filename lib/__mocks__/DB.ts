const DB = jest.genMockFromModule("../DB").default;

export function mockRows(rows = [], { once = true } = {}): void {
	const result = { rows, rowCount: rows.length };
	if (once) {
		DB.query.mockResolvedValueOnce(result);
	} else {
		DB.query.mockResolvedValue(result);
	}
}

DB.__mockRows = mockRows;
DB.__mockRows([], { once: false });

export default DB;
