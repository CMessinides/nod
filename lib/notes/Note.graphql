type Note {
	id: ID!
	title: String!
	slug: String!
	createdAt: Date!
	modifiedAt: Date!
	notebook: Notebook!
	content: [NoteChunk]!
}
