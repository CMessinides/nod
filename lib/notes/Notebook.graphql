type Notebook {
	id: ID!
	title: String!
	slug: String!
	description: String
	createdAt: Date!
	notes: [Note]!
}
