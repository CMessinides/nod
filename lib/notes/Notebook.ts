import { NotebookModel } from "./interfaces";
import DB from "../DB";
import pipe from "lodash/fp/pipe";
import { makeSluggable } from "../slugs";
import Note from "./Note";

interface NotebookRecord {
	id: number;
	title: string;
	description: string;
	created_at: Date;
}

type NotebookParams = Omit<NotebookModel, "notes" | "slug">;

const prepareNotebook: (notebook: Notebook) => Notebook = pipe(makeSluggable);

export default class Notebook implements NotebookModel {
	id: number;
	title: string;
	slug: string;
	description: string;
	createdAt: Date;

	static async findById(id: number): Promise<Notebook> {
		const { rows, rowCount } = await DB.query(
			"SELECT * FROM notebooks WHERE id = $1 LIMIT 1",
			[id]
		);

		if (rowCount === 0) {
			return null;
		}

		return Notebook.fromRecord(rows[0] as NotebookRecord);
	}

	static fromRecord({ created_at, ...props }: NotebookRecord): Notebook {
		return new Notebook({ ...props, createdAt: created_at });
	}

	constructor(params: NotebookParams) {
		Object.assign(this, params);
		prepareNotebook(this);
	}

	notes(): Promise<Note[]> {
		return Note.getAllInNotebook(this.id);
	}
}
