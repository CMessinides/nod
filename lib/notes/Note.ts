import { NoteModel, NoteChunkType } from "./interfaces";
import DB from "../DB";
import { makeSluggable } from "../slugs";
import pipe from "lodash/fp/pipe";
import sortBy from "lodash/fp/sortBy";
import reverse from "lodash/fp/reverse";
import Notebook from "./Notebook";
import LinkedList from "../LinkedList";
import NoteTaskListChunk, {
	NoteTaskListChunkRecord
} from "./NoteTaskListChunk";
import NoteTextContentChunk, {
	NoteTextContentChunkRecord
} from "./NoteTextContentChunk";
import { NoteChunkRecord } from "./NoteContentChunk";

interface NoteRecord {
	id: number;
	title: string;
	created_at: Date;
	modified_at: Date;
	notebook_id: number;
}

type NoteParams = Omit<NoteModel, "notebook" | "content" | "slug"> & {
	notebookId: number;
};

const prepareNote: (note: Note) => Note = pipe(makeSluggable);
const sortByModifiedDate = pipe(
	sortBy<Note>("modifiedAt"),
	reverse
);

type NoteContentChunk = NoteTaskListChunk | NoteTextContentChunk;
function modelChunkByType(record: NoteChunkRecord): NoteContentChunk {
	switch (record.type) {
		case NoteChunkType.TASK_LIST:
			return NoteTaskListChunk.fromRecord(record as NoteTaskListChunkRecord);
		case NoteChunkType.TEXT_CONTENT:
			return NoteTextContentChunk.fromRecord(
				record as NoteTextContentChunkRecord
			);
		default:
			throw new Error("Unrecognized content chunk type: " + record.type);
	}
}

export default class Note implements NoteModel {
	id: number;
	title: string;
	slug: string;
	createdAt: Date;
	modifiedAt: Date;
	private notebookId: number;

	static async findById(id: number): Promise<Note> {
		const { rows, rowCount } = await DB.query(
			"SELECT * FROM notes WHERE id = $1 LIMIT 1",
			[id]
		);

		if (rowCount === 0) {
			return null;
		}

		return Note.fromRecord(rows[0] as NoteRecord);
	}

	static async getAllInNotebook(notebookId: number): Promise<Note[]> {
		const { rows }: { rows: NoteRecord[] } = await DB.query(
			"SELECT * FROM notes WHERE notebook_id = $1",
			[notebookId]
		);

		return sortByModifiedDate(rows.map(Note.fromRecord));
	}

	static fromRecord({
		id,
		title,
		created_at,
		modified_at,
		notebook_id
	}: NoteRecord): Note {
		return new Note({
			id,
			title,
			createdAt: created_at,
			modifiedAt: modified_at,
			notebookId: notebook_id
		});
	}

	constructor(params: NoteParams) {
		for (const key of Object.keys(params)) {
			this[key] = params[key];
		}

		prepareNote(this);
	}

	notebook(): Promise<Notebook> {
		return Notebook.findById(this.notebookId);
	}

	async content(): Promise<NoteContentChunk[]> {
		const { rows }: { rows: NoteChunkRecord[] } = await DB.query(
			"SELECT note_chunks.id, prev_chunk_id, type, name, list_id, text FROM note_chunks FULL JOIN note_text_content_chunks ON note_chunks.id = note_text_content_chunks.id FULL JOIN (SELECT note_task_list_chunks.id, list_id, name FROM note_task_list_chunks JOIN task_lists ON note_task_list_chunks.list_id = task_lists.id) AS full_note_task_list_chunks ON note_chunks.id = full_note_task_list_chunks.id WHERE note_id = $1",
			[this.id]
		);

		return Array.from(LinkedList(rows, "id", "prev_chunk_id")).map(
			modelChunkByType
		);
	}
}
