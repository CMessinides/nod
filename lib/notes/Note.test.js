import Note from "./Note";

// For mocking
jest.mock("../DB");
import DB from "../DB";
jest.mock("./Notebook");
import Notebook from "./Notebook";
import MockRecords from "./fixtures/NoteRecords";
import MockChunkRecords from "./fixtures/NoteContentChunkRecords";

describe("static methods", () => {
	describe("findById", () => {
		it("should return the note with the given ID", () => {
			const record = MockRecords[0];
			const { id } = record;
			DB.__mockRows([record]);

			expect(Note.findById(id)).resolves.toMatchSnapshot();
		});

		it("should return null if the note does not exist", () => {
			expect(Note.findById(2)).resolves.toBe(null);
		});
	});

	describe("getAllInNotebook", () => {
		it("should return the notes in the given notebook sorted by modified date", () => {
			const notebookId = 1;
			const records = MockRecords.filter(
				record => record.notebook_id === notebookId
			);
			DB.__mockRows(records);

			expect(Note.getAllInNotebook(notebookId)).resolves.toMatchSnapshot();
		});

		it("should return an empty array if no notes exist", () => {
			expect(Note.getAllInNotebook(2)).resolves.toHaveLength(0);
		});
	});
});

describe("instance methods", () => {
	describe("notebook", () => {
		it("should return the note's parent notebook", () => {
			const note = Note.fromRecord(MockRecords[0]);
			const notebook = new Notebook({ id: note.notebookId });
			Notebook.findById.mockResolvedValueOnce(notebook);

			expect(note.notebook()).resolves.toBe(notebook);
		});
	});

	describe("content", () => {
		it("should return the note's content chunks in order", () => {
			const note = Note.fromRecord(MockRecords[0]);
			const chunkRecords = MockChunkRecords.filter(
				record => record.note_id === note.id
			);
			DB.__mockRows(chunkRecords);

			expect(note.content()).resolves.toMatchSnapshot();
		});
	});
});
