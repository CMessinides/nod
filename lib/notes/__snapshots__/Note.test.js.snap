// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`instance methods content should return the note's content chunks in order 1`] = `
Array [
  NoteTaskListChunk {
    "id": 1,
    "listId": 46,
    "name": "Outdoors",
    "type": "TASK_LIST",
  },
  NoteTextContentChunk {
    "id": 2,
    "text": "Proin leo odio, porttitor id, consequat in, consequat ut, nulla. Sed accumsan felis. Ut at dolor quis odio consequat varius. Integer ac leo. Pellentesque ultrices mattis odio.",
    "type": "TEXT_CONTENT",
  },
  NoteTaskListChunk {
    "id": 3,
    "listId": 39,
    "name": "Tools",
    "type": "TASK_LIST",
  },
]
`;

exports[`static methods findById should return the note with the given ID 1`] = `
Note {
  "createdAt": 1970-01-18T12:32:29.373Z,
  "id": 1,
  "modifiedAt": 1970-01-18T22:40:08.092Z,
  "notebookId": 2,
  "slug": "gekko-gecko--1",
  "title": "Gekko gecko",
}
`;

exports[`static methods getAllInNotebook should return the notes in the given notebook sorted by modified date 1`] = `
Array [
  Note {
    "createdAt": 1970-01-18T13:29:58.688Z,
    "id": 7,
    "modifiedAt": 1970-01-19T02:26:11.161Z,
    "notebookId": 1,
    "slug": "columba-livia--7",
    "title": "Columba livia",
  },
  Note {
    "createdAt": 1970-01-18T15:30:25.338Z,
    "id": 2,
    "modifiedAt": 1970-01-19T02:12:06.287Z,
    "notebookId": 1,
    "slug": "ardea-cinerea--2",
    "title": "Ardea cinerea",
  },
]
`;
